Make IPC::Connection::Handle move-only.

mattwoodrow opened the pull request:
#### b44aa50aceaea3c7b7e4cf783c43ed93e4a4047d
<pre>
Make IPC::Connection::Handle move-only.
<a href="https://bugs.webkit.org/show_bug.cgi?id=257579">https://bugs.webkit.org/show_bug.cgi?id=257579</a>

Reviewed by Dan Glastonbury.

Connection::Handle represents a way to construct the remote end of a connection, and its lifetime controls disconnection messages being sent to the local end.

We shouldn&apos;t ever want to copy these, only one remote end can be constructed.
t show
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::Handle::encode):
(IPC::Connection::Handle::decode):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::Handle::Handle):
(IPC::Connection::Handle::operator bool const):
(IPC::Connection::Identifier::Identifier):
(IPC::Connection::Identifier::operator bool const):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::getNetworkProcessConnection):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::create):
* Source/WebKit/WebProcess/Inspector/WebInspectorUI.cpp:
(WebKit::WebInspectorUI::updateConnection):
* Source/WebKit/WebProcess/Network/NetworkProcessConnectionInfo.h:
(WebKit::NetworkProcessConnectionInfo::encode):
(WebKit::NetworkProcessConnectionInfo::encode const): Deleted.
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::MockTestMessageWithConnection::arguments):
(TestWebKitAPI::MockTestMessageWithConnection::MockTestMessageWithConnection):
(TestWebKitAPI::TEST_F):
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.cpp:
(TestWebKitAPI::ConnectionTestBase::setupBase):

Canonical link: <a href="https://commits.webkit.org/264928@main">https://commits.webkit.org/264928@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d7cb1ad3af20787d99cc08b7bd4ef8020099ed6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9172 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9453 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9681 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10828 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9180 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11441 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9420 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11957 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9319 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10268 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10984 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/9295 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7563 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8370 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15837 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/8667 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8517 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11836 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9019 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7357 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8248 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2218 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/12471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8778 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -177,19 +177,48 @@ class Connection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Connec
+        Handle(MachSendRight&& sendRight)
+            : handle(WTFMove(sendRight))
+        { }
+        explicit operator bool() const { return MACH_PORT_VALID(handle.sendRight()); }
```
`MachSendRight` has `operator bool` that only checks half of what `MACH_PORT_VALID` checks. (`MachSendRight::operator bool()` only checks against `MACH_PORT_NULL`)

Should `MachSendRight::operator bool()` be updated to use `MACH_PORT_VALID` and this code can be `explicit operator bool() const { return !!handle; }` in all cases?

mattwoodrow replied:
`MachSendRight` has explicit code for handing the 'dead name' value (the other half of MACH_PORT_VALID), including refcounting it. That said, I'm not sure if any consumer actually wants that behaviour.

djg commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -211,7 +240,7 @@ class Connection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Connec
 #elif OS(DARWIN)
         explicit Identifier(Handle&& handle)
-            : Identifier(handle.leakSendRight())
+            : Identifier(handle.handle.leakSendRight())
```
handle.release/handle.leak/handle.leakSendRight 🤦

mattwoodrow replied:
Sad indeed. Ideally we'd have move-only semantics for MachSendRight (bug 257580), and a common interface for the darwin/win32/posix implementations, and then this wrapper class can go away.

djg reviewed (approved):
:shipit:

The change to `MachSendRight` isn't necessary, but might be a good follow up.

mattwoodrow committed:
Make IPC::Connection::Handle move-only.
https://bugs.webkit.org/show_bug.cgi?id=257579

Reviewed by Dan Glastonbury.

Connection::Handle represents a way to construct the remote end of a connection, and its lifetime controls disconnection messages being sent to the local end.

We shouldn't ever want to copy these, only one remote end can be constructed.
t show
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::Handle::encode):
(IPC::Connection::Handle::decode):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::Handle::Handle):
(IPC::Connection::Handle::operator bool const):
(IPC::Connection::Identifier::Identifier):
(IPC::Connection::Identifier::operator bool const):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::getNetworkProcessConnection):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::create):
* Source/WebKit/WebProcess/Inspector/WebInspectorUI.cpp:
(WebKit::WebInspectorUI::updateConnection):
* Source/WebKit/WebProcess/Network/NetworkProcessConnectionInfo.h:
(WebKit::NetworkProcessConnectionInfo::encode):
(WebKit::NetworkProcessConnectionInfo::encode const): Deleted.
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::MockTestMessageWithConnection::arguments):
(TestWebKitAPI::MockTestMessageWithConnection::MockTestMessageWithConnection):
(TestWebKitAPI::TEST_F):
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.cpp:
(TestWebKitAPI::ConnectionTestBase::setupBase):

Canonical link: https://commits.webkit.org/264928@main
